pkg/infrastructure/powervs/clusterapi/powervs.go

```diff
@@ -12,15 +23,240 @@ type Provider struct {
+		switch {
+		case lbExtExp.MatchString(lbKey):
+			prefix = "api."
+		case lbIntExp.MatchString(lbKey):
+			prefix = "api-int."
+		}
+		logrus.Debugf("prefix = %s", prefix)
+
+		hostname = fmt.Sprintf("%s%s", prefix, infraID)
+
+		logrus.Debugf("InfraReady: crn = %s, base domain = %s, hostname = %s, cname = %s",
+			instanceCRN,
+			in.InstallConfig.PowerVS.BaseDomain,
+			hostname,
+			*loadBalancerStatus.Hostname)
+
+		err = client.CreateDNSRecord(ctx,
+			instanceCRN,
+			in.InstallConfig.PowerVS.BaseDomain,
+			hostname,
+			*loadBalancerStatus.Hostname)
+		if err != nil {
+			return fmt.Errorf("failed to create a DNS CNAME record (%s, %s): %w",
+				hostname,
+				*loadBalancerStatus.Hostname,
+				err)
+		}
+	}
+
+	// Step 2.
+	// Add to security group rules
+	for _, port = range ports {
+		rule = &vpcv1.SecurityGroupRulePrototype{
+			Direction: ptr.To("inbound"),
+			Protocol:  ptr.To("tcp"),
+			PortMin:   ptr.To(port),
+			PortMax:   ptr.To(port),
+		}
+
+		err = client.AddSecurityGroupRule(ctx, *powerVSCluster.Status.VPC.ID, rule)
+		if err != nil {
+			return fmt.Errorf("failed to add security group rule for port %d: %w", port, err)
+		}
+	}
+
+	rule = &vpcv1.SecurityGroupRulePrototype{
+		Direction: ptr.To("inbound"),
+		Protocol:  ptr.To("icmp"),
+	}
+
+	err = client.AddSecurityGroupRule(ctx, *powerVSCluster.Status.VPC.ID, rule)
+	if err != nil {
+		return fmt.Errorf("failed to add security group rule for icmp: %w", err)
+	}
+
+	return nil
+}
+
+// PostProvision should be called to add or update PowerVS resources after provisioning has completed.
+func (p Provider) PostProvision(ctx context.Context, in clusterapi.PostProvisionInput) error {
+	var (
+		client             *powervsconfig.Client
+		refServiceInstance *capibm.IBMPowerVSResourceReference
+		sshKeyName         string
+		err                error
+	)
+
+	// SAD: client in the Metadata struct is lowercase and therefore private
+	// client = in.InstallConfig.PowerVS.client
+	client, err = powervsconfig.NewClient()
+	if err != nil {
+		return fmt.Errorf("failed to get NewClient in PostProvision: %w", err)
+	}
+	logrus.Debugf("PostProvision: NewClient returns %+v", client)
+
+	// Step 1.
+	// Create worker ssh key
+	// Get the bootstrap machine from the provider
+	key := crclient.ObjectKey{
+		Name:      fmt.Sprintf("%s-bootstrap", in.InfraID),
+		Namespace: capiutils.Namespace,
+	}
+	logrus.Debugf("PostProvision: machine key = %+v", key)
+	powerVSMachine := &capibm.IBMPowerVSMachine{}
+
+	err = in.Client.Get(ctx, key, powerVSMachine)
+	if err != nil {
+		return fmt.Errorf("failed to get PowerVS machine in PostProvision: %w", err)
+	}
+	logrus.Debugf("PostProvision: machine = %+v", powerVSMachine)
+	logrus.Debugf("PostProvision: machine.Spec.ServiceInstance = %+v", powerVSMachine.Spec.ServiceInstance)
+	refServiceInstance = powerVSMachine.Spec.ServiceInstance
+
+	if in.InstallConfig.Config.SSHKey == "" {
+		return fmt.Errorf("install config's ssh key is empty?")
+	}
```

Comment:
If it's a requirement, it should be checked during installconfig validation?

Reply:
I will add a TODO comment entry